clickhouse/columns/array.h

```diff
@@ -38,19 +53,209 @@ class ColumnArray : public Column {
+protected:
+    template<typename T> friend class ColumnArrayT;
+
+    ColumnArray(ColumnArray&& array);
+    struct ShareOwnershipTag {};
+    ColumnArray(ColumnRef data, ShareOwnershipTag tag);
 
+    size_t GetOffset(size_t n) const;
     size_t GetSize(size_t n) const;
+    ColumnRef GetData();
+    void AddOffset(size_t n);
+    void Reset();
 
 private:
     ColumnRef data_;
     std::shared_ptr<ColumnUInt64> offsets_;
 };
 
+template <typename NestedColumnType>
+class ColumnArrayT : public ColumnArray {
+public:
+    class ArrayWrapper;
+    using ValueType = ArrayWrapper;
+
+    explicit ColumnArrayT(std::shared_ptr<NestedColumnType> data)
+        : ColumnArray(data, ColumnArray::ShareOwnershipTag{})
+        , typed_nested_data_(data)
+    {}
+
+    template <typename ...Args>
+    explicit ColumnArrayT(Args &&... args)
+        : ColumnArrayT(std::make_shared<NestedColumnType>(std::forward<Args>(args)...))
+    {}
+
+    // Helper to allow wrapping a "typeless" ColumnArray
+    explicit ColumnArrayT(ColumnArray&& array)
+        : ColumnArray(std::move(array))
+        , typed_nested_data_(this->GetData()->template AsStrict<NestedColumnType>())
+    {}
+
+    /** Create a ColumnArrayT from a ColumnArray, without copying data and offsets.
+     *  Ownership of column internals is transferred to returned object, original (argument) object
+     *  MUST NOT BE USED IN ANY WAY, it is only safe to dispose it.
+     *
+     *  Throws an exception of `col` is of wrong type, it is safe to use original col in this case.
+     *  This is a static method to make such conversion verbose.
+     */
+    static auto Wrap(ColumnArray&& col) {
+        if constexpr (std::is_base_of_v<ColumnArray, NestedColumnType> && !std::is_same_v<ColumnArray, NestedColumnType>) {
+            // assuming NestedColumnType is ArrayT specialization
+
+            auto result = std::make_shared<ColumnArrayT<NestedColumnType>>(NestedColumnType::Wrap(col.GetData()));
+            result->offsets_ = col.offsets_;
+
+            return result;
+        } else {
+            return std::make_shared<ColumnArrayT<NestedColumnType>>(std::move(col));
+        }
+    }
+
+    static auto Wrap(Column&& col) {
+        return Wrap(std::move(dynamic_cast<ColumnArray&&>(col)));
+    }
+
+    // Helper to simplify integration with other APIs
+    static auto Wrap(ColumnRef&& col) {
+        return Wrap(std::move(*col->AsStrict<ColumnArray>()));
+    }
+
+    class ArrayWrapper {
+        const std::shared_ptr<NestedColumnType> typed_nested_data_;
+        const size_t offset_;
+        const size_t size_;
+
+    public:
+        using ValueType = typename NestedColumnType::ValueType;
+
+        ArrayWrapper(std::shared_ptr<NestedColumnType> data, size_t offset = 0, size_t size = std::numeric_limits<size_t>::max())
+            : typed_nested_data_(data)
+            , offset_(offset)
+            , size_(std::min(typed_nested_data_->Size() - offset, size))
```

Comment:
If `size` is bigger than `typed_nested_data_->Size() - offset`, I assume the client would expect to get `typed_nested_data_->Size() - offset` of size. If that's the case, then it's fine. Otherwise, maybe we should throw an exception.

Reply:
well, we build `ArrayWrapper` ourselves, and IMO such situation would never happen.